<a href="https://colab.research.google.com/github/cagBRT/PerformanceEnhancement/blob/main/Pandas_Performance_Enhancement.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

This notebook covers some techniques for speeding up performance for large datasets (>= 1 million rows) using Pandas and Numpy

In [1]:
# Clone the entire repo.
!git clone -l -s https://github.com/cagBRT/Intro-to-Pandas.git cloned-repo
%cd cloned-repo
!ls

Cloning into 'cloned-repo'...
remote: Enumerating objects: 243, done.
remote: Counting objects: 100% (98/98), done.
remote: Compressing objects: 100% (98/98), done.
remote: Total 243 (delta 60), reused 0 (delta 0), pack-reused 145
Receiving objects: 100% (243/243), 6.35 MiB | 4.63 MiB/s, done.
Resolving deltas: 100% (130/130), done.
/content/cloned-repo
 adult.csv	        P2_Pandas_Data_Prep.ipynb	       titanic.csv
 deniro.csv	        P3_Pandas.ipynb			       train.csv
 letter_frequency.csv  'Pandas(1).png'			       trainTitanic.csv
 oscarNoErrors.csv      Pandas_Performance_Enhancement.ipynb   UFO.csv
 oscar_winners.csv      README.md			       xP1_Intro_to_Pandas.ipynb
 P1_Pandas_ES.ipynb     test.csv
 P1_Pandas.ipynb        testTitanic.csv


## Get the data

1. Download the dataset here:
###https://www.kaggle.com/competitions/tabular-playground-series-sep-2021/rules

This dataset is 1GB <br>
It has nearly 1million rows and 128 columns

2. It will be be in zip format. <br>

In [3]:
import zipfile

Upload to the local directory:<br>   

*   tablular-playground-series-sep-2021


This can take up to 20 minutes to load

**You can continue with the notebook, this file is not needed for a little while.**



---



---



In [4]:
import pandas as pd
import numpy as np

## Use replace to replace specific values in Pandas datasets

Speed is the first benefit of replace, <br>
the second benefit is its flexibility.

We can replace all question marks with NaN - an operation that would take multiple calls with index-based replacement.

Nested replacement helps when you only want to affect the values of specific columns. Here, we are replacing values only in education and income columns.



In [5]:
s = pd.Series([1, 2, 3, 4, 5])
s

0    1
1    2
2    3
3    4
4    5
dtype: int64

In [6]:
s.replace(1, 5)

0    5
1    2
2    3
3    4
4    5
dtype: int64

In [7]:
s.replace([1, 2], method='bfill')

0    3
1    3
2    3
3    4
4    5
dtype: int64

In [8]:
df = pd.DataFrame({'A': [0, 1, 2, 3, 4],
                   'B': [5, 6, 7, 8, 9],
                   'C': ['a', 'b', 'c', 'd', 'e']})
df.replace(0, 5)

A  B  C
0  5  5  a
1  1  6  b
2  2  7  c
3  3  8  d
4  4  9  e

In [9]:
df.replace([0, 1, 2, 3], 4)

A  B  C
0  4  5  a
1  4  6  b
2  4  7  c
3  4  8  d
4  4  9  e

In [10]:
df.replace([0, 1, 2, 3], [4, 3, 2, 1])

A  B  C
0  4  5  a
1  3  6  b
2  2  7  c
3  1  8  d
4  4  9  e

In [11]:
adult_income = pd.read_csv("adult.csv")

In [12]:
adult_income.shape

(48842, 15)

In [13]:
adult_income.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'educational-num',
       'marital-status', 'occupation', 'relationship', 'race', 'gender',
       'capital-gain', 'capital-loss', 'hours-per-week', 'native-country',
       'income'],
      dtype='object')

In [14]:
adult_income.isna().sum()

age                0
workclass          0
fnlwgt             0
education          0
educational-num    0
marital-status     0
occupation         0
relationship       0
race               0
gender             0
capital-gain       0
capital-loss       0
hours-per-week     0
native-country     0
income             0
dtype: int64

In [15]:
adult_income.replace(to_replace="?", value=np.nan, inplace=True)

In [16]:
adult_income.isna().sum()

age                   0
workclass          2799
fnlwgt                0
education             0
educational-num       0
marital-status        0
occupation         2809
relationship          0
race                  0
gender                0
capital-gain          0
capital-loss          0
hours-per-week        0
native-country      857
income                0
dtype: int64

replace allows using lists or dictionaries to change multiple values simultaneously:



In [17]:
adult_income.gender.value_counts()

Male      32650
Female    16192
Name: gender, dtype: int64

In [18]:
adult_income.replace(["Male", "Female"], ["M", "F"], inplace=True)

In [19]:
adult_income.gender.value_counts()

M    32650
F    16192
Name: gender, dtype: int64

When replacing a list of values with another, they will have a one-to-one, index-to-index mapping.

In [20]:
adult_income["native-country"].value_counts()

United-States                 43832
Mexico                          951
Philippines                     295
Germany                         206
Puerto-Rico                     184
Canada                          182
El-Salvador                     155
India                           151
Cuba                            138
England                         127
China                           122
South                           115
Jamaica                         106
Italy                           105
Dominican-Republic              103
Japan                            92
Guatemala                        88
Poland                           87
Vietnam                          86
Columbia                         85
Haiti                            75
Portugal                         67
Taiwan                           65
Iran                             59
Greece                           49
Nicaragua                        49
Peru                             46
Ecuador                     

In [21]:
adult_income.replace({"United States": "USA", "US": "USA"}, inplace=True)

In [22]:
adult_income["native-country"].value_counts()

United-States                 43832
Mexico                          951
Philippines                     295
Germany                         206
Puerto-Rico                     184
Canada                          182
El-Salvador                     155
India                           151
Cuba                            138
England                         127
China                           122
South                           115
Jamaica                         106
Italy                           105
Dominican-Republic              103
Japan                            92
Guatemala                        88
Poland                           87
Vietnam                          86
Columbia                         85
Haiti                            75
Portugal                         67
Taiwan                           65
Iran                             59
Greece                           49
Nicaragua                        49
Peru                             46
Ecuador                     

In [23]:
adult_income.replace({"United States": "USA", "US": "USA"}, inplace=True)

In [24]:
adult_income["native-country"].value_counts()

United-States                 43832
Mexico                          951
Philippines                     295
Germany                         206
Puerto-Rico                     184
Canada                          182
El-Salvador                     155
India                           151
Cuba                            138
England                         127
China                           122
South                           115
Jamaica                         106
Italy                           105
Dominican-Republic              103
Japan                            92
Guatemala                        88
Poland                           87
Vietnam                          86
Columbia                         85
Haiti                            75
Portugal                         67
Taiwan                           65
Iran                             59
Greece                           49
Nicaragua                        49
Peru                             46
Ecuador                     

In [25]:
adult_income.education.value_counts()

HS-grad         15784
Some-college    10878
Bachelors        8025
Masters          2657
Assoc-voc        2061
11th             1812
Assoc-acdm       1601
10th             1389
7th-8th           955
Prof-school       834
9th               756
12th              657
Doctorate         594
5th-6th           509
1st-4th           247
Preschool          83
Name: education, dtype: int64

In [26]:
adult_income.income.value_counts()

<=50K    37155
>50K     11687
Name: income, dtype: int64

In [27]:
adult_income.replace(
    {
        "education": {"HS-grad": "High school", "Some-college": "College", "12th":"High school"},
        "income": {"<=50K": 0, ">50K": 1},
    },
    inplace=True,
)

In [28]:
adult_income.education.value_counts()

High school    16441
College        10878
Bachelors       8025
Masters         2657
Assoc-voc       2061
11th            1812
Assoc-acdm      1601
10th            1389
7th-8th          955
Prof-school      834
9th              756
Doctorate        594
5th-6th          509
1st-4th          247
Preschool         83
Name: education, dtype: int64

In [29]:
adult_income.income.value_counts()

0    37155
1    11687
Name: income, dtype: int64

# iloc vs loc

For choosing a row or multiple rows, iloc is faster

In [30]:
%time adult_income.iloc[range(10000)]
%time adult_income.loc[range(10000)]

CPU times: user 4.08 ms, sys: 873 µs, total: 4.95 ms
Wall time: 5.51 ms
CPU times: user 3.55 ms, sys: 0 ns, total: 3.55 ms
Wall time: 3.4 ms


age         workclass  fnlwgt    education  educational-num  \
0      25           Private  226802         11th                7   
1      38           Private   89814  High school                9   
2      28         Local-gov  336951   Assoc-acdm               12   
3      44           Private  160323      College               10   
4      18               NaN  103497      College               10   
...   ...               ...     ...          ...              ...   
9995   66  Self-emp-not-inc  176315    Bachelors               13   
9996   35           Private  187167  High school                9   
9997   24           Private  241582      College               10   
9998   31           Private  247328         11th                7   
9999   75  Self-emp-not-inc  157778    Assoc-voc               11   

          marital-status         occupation   relationship   race gender  \
0          Never-married  Machine-op-inspct      Own-child  Black      M   
1     Married-civ-spouse    Farming-fishing        Husband  White      M   
2     Married-civ-spouse    Protective-serv        Husband  White      M   
3     Married-civ-spouse  Machine-op-inspct        Husband  Black      M   
4          Never-married                NaN      Own-child  White      F   
...                  ...                ...            ...    ...    ...   
9995            Divorced              Sales  Not-in-family  White      M   
9996       Never-married       Adm-clerical      Unmarried  White      F   
9997       Never-married              Sales  Not-in-family  White      M   
9998  Married-civ-spouse    Protective-serv        Husband  White      M   
9999            Divorced     Prof-specialty  Not-in-family  White      F   

      capital-gain  capital-loss  hours-per-week native-country  income  
0                0             0              40  United-States       0  
1                0             0              50  United-States       0  
2                0             0              40  United-States       1  
3             7688             0              40  United-States       1  
4                0             0              30  United-States       0  
...            ...           ...             ...            ...     ...  
9995           401             0              20  United-States       0  
9996             0             0              40  United-States       0  
9997             0             0              33  United-States       0  
9998             0             0              40  United-States       0  
9999         15831             0              50  United-States       1  

[10000 rows x 15 columns]

Choosing columns:

In [31]:
%time adult_income.loc[:,["workclass","occupation","gender"]]
%time adult_income.iloc[:,[1,6,9]]

CPU times: user 2.57 ms, sys: 79 µs, total: 2.65 ms
Wall time: 4.07 ms
CPU times: user 1.71 ms, sys: 0 ns, total: 1.71 ms
Wall time: 1.55 ms


workclass         occupation gender
0           Private  Machine-op-inspct      M
1           Private    Farming-fishing      M
2         Local-gov    Protective-serv      M
3           Private  Machine-op-inspct      M
4               NaN                NaN      F
...             ...                ...    ...
48837       Private       Tech-support      F
48838       Private  Machine-op-inspct      M
48839       Private       Adm-clerical      F
48840       Private       Adm-clerical      M
48841  Self-emp-inc    Exec-managerial      F

[48842 rows x 3 columns]

In [32]:
%time adult_income.sample(7,axis=0)

CPU times: user 2.17 ms, sys: 27 µs, total: 2.2 ms
Wall time: 2.21 ms


age         workclass  fnlwgt    education  educational-num  \
35816   27  Self-emp-not-inc  208577  High school                9   
3775    43           Private  103759  High school                9   
47057   17               NaN  219277         11th                7   
48609   31           Private  254304         10th                6   
16796   57               NaN  222216    Assoc-voc               11   
39328   25           Private  390316          9th                5   
13880   54           Private  249352    Bachelors               13   

           marital-status        occupation   relationship   race gender  \
35816       Never-married      Craft-repair  Not-in-family  White      M   
3775   Married-civ-spouse  Transport-moving        Husband  White      M   
47057       Never-married               NaN      Own-child  White      F   
48609            Divorced      Craft-repair  Not-in-family  White      M   
16796             Widowed               NaN      Unmarried  White      F   
39328  Married-civ-spouse      Craft-repair        Husband  White      M   
13880  Married-civ-spouse      Craft-repair        Husband  Black      M   

       capital-gain  capital-loss  hours-per-week native-country  income  
35816             0          2258              50  United-States       0  
3775              0             0              40  United-States       0  
47057             0             0              20  United-States       0  
48609             0             0              38  United-States       0  
16796             0             0              38  United-States       0  
39328             0             0              40         Mexico       0  
13880             0             0              40  United-States       0

In [33]:
%time adult_income.sample(5, axis=1).sample(7, axis=0)

CPU times: user 3.02 ms, sys: 993 µs, total: 4.01 ms
Wall time: 4.03 ms


educational-num  fnlwgt  capital-loss native-country         occupation
8654                16  443508             0  United-States    Exec-managerial
12559               13  140676             0  United-States              Sales
13938               10  336162             0  United-States       Adm-clerical
27538               14  372525             0  United-States     Prof-specialty
42193                4  236012             0  United-States  Handlers-cleaners
20945                9  183970             0  United-States  Handlers-cleaners
5069                 9  219751             0  United-States    Exec-managerial

# Iterating efficiently

**The golden rule for applying operations on entire columns or data frames is to never use loops**

Think about arrays as vectors and the whole data frame as a matrix

If you want to perform any mathematical operation on one or more columns, there is a good chance that the operation is vectorized in Pandas.

For example, the built-in Python operators like +, -, *, /, ** work just like on vectors.

To get a taste of vectorization, let’s perform some operations on a massive dataset. We will choose ~1M row dataset of the old Kaggle TPS September competition:

# Datatable

Datatable is a python library for manipulating tabular data.


It supports out-of-memory datasets, multi-threaded data processing, and flexible API.



The datatable module emphasizes speed and big data support (an area that pandas struggles with); it also has an expressive and concise syntax, which makes datatable also useful for small datasets.

Note: in pandas, there are two fundamental data structures: Series and DataFrame.

In [34]:
!python3 -m pip install -U pip
!python3 -m pip install -U datatable

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.1/2.1 MB 16.3 MB/s eta 0:00:00
  Attempting uninstall: pip
    Found existing installation: pip 23.1.2
    Uninstalling pip-23.1.2:
      Successfully uninstalled pip-23.1.2
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 82.0/82.0 MB 6.0 MB/s eta 0:00:00


In [35]:
import datatable as dt

**Only do this part if the tabular-playground-series-sssep-2021 is done uploading.**<br>

In [37]:
!unzip "/content/tabular-playground-series-sep-2021 (3).zip"

Archive:  /content/tabular-playground-series-sep-2021 (3).zip
  inflating: sample_solution.csv     
replace test.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: A
  inflating: test.csv                
  inflating: train.csv               


In [38]:
tps = dt.fread("/content/cloned-repo/train.csv").to_pandas()
tps.shape

(957919, 120)

The shape of train.csv should be (957919, 120)

## Apply

The fastest built-in iterator of Pandas is apply.

Apply is about 600 times faster than the fastest iterator. But we can do even better — vectorization is even faster when used on NumPy arrays:

Add .values to get the underlying NumPy ndarray of Pandas Series.

NumPy arrays are faster because they don't perform additional calls for indexing and data type

In [39]:
def crazy_function(col1, col2, col3):
    return np.sqrt(col1 ** 3 + col2 ** 2 + col3 * 10)

Time the crazy_function on three columns using apply

In [40]:
%time tps['f1000'] = tps.apply(lambda row: crazy_function(row['f1'], row['f56'], row['f44']), axis=1)

<ipython-input-39-c220950852c4>:2: RuntimeWarning: invalid value encountered in sqrt
  return np.sqrt(col1 ** 3 + col2 ** 2 + col3 * 10)


CPU times: user 16.1 s, sys: 3.56 s, total: 19.7 s
Wall time: 19.7 s


## Vectors

Watch what happens when we pass columns as vectors rather than scalars. No need to modify the function:

In [41]:
%time tps['f1001'] = crazy_function(tps['f1'], tps['f56'], tps['f44'])

CPU times: user 43.3 ms, sys: 772 µs, total: 44.1 ms
Wall time: 52.8 ms


/usr/local/lib/python3.10/dist-packages/pandas/core/arraylike.py:402: RuntimeWarning: invalid value encountered in sqrt
  result = getattr(ufunc, method)(*inputs, **kwargs)


Using apply: CPU time: 19.7 s<br>
Using vectors: CPU time: 52.8 ms<br>

Vectors much faster that using apply!

## Vectors with Numpy

In [ ]:
%time tps['f1001'] = crazy_function(tps['f1'].values, tps['f56'].values, tps['f44'].values)

CPU times: user 51.5 ms, sys: 0 ns, total: 51.5 ms
Wall time: 65.7 ms


<ipython-input-37-c220950852c4>:2: RuntimeWarning: invalid value encountered in power
  return np.sqrt(col1 ** 3 + col2 ** 2 + col3 * 10)
<ipython-input-37-c220950852c4>:2: RuntimeWarning: invalid value encountered in square
  return np.sqrt(col1 ** 3 + col2 ** 2 + col3 * 10)
<ipython-input-37-c220950852c4>:2: RuntimeWarning: invalid value encountered in multiply
  return np.sqrt(col1 ** 3 + col2 ** 2 + col3 * 10)
<ipython-input-37-c220950852c4>:2: RuntimeWarning: invalid value encountered in sqrt
  return np.sqrt(col1 ** 3 + col2 ** 2 + col3 * 10)


Pandas has a few more tricks up its sleeve.

**Fair warning, though — these won’t benefit you much unless you have upwards of +1M rows.**

In [42]:
massive_df = pd.concat([tps.drop(["f1000", "f1001"], axis=1)] * 10)
massive_df.shape

(9579190, 120)

In [43]:
memory_usage = massive_df.memory_usage(deep=True)
memory_usage_in_mbs = np.sum(memory_usage / 1024 ** 2)
memory_usage_in_mbs

8742.604093551636

Use our crazy_function, start with NumPy vectorization as a baseline


It takes about 0.3 seconds for a 10M row dataset

In [44]:
%%time

massive_df["f1001"] = crazy_function(
    massive_df["f1"].values, massive_df["f56"].values, massive_df["f44"].values
)

CPU times: user 393 ms, sys: 69.8 ms, total: 463 ms
Wall time: 490 ms


<ipython-input-39-c220950852c4>:2: RuntimeWarning: invalid value encountered in power
  return np.sqrt(col1 ** 3 + col2 ** 2 + col3 * 10)
<ipython-input-39-c220950852c4>:2: RuntimeWarning: invalid value encountered in square
  return np.sqrt(col1 ** 3 + col2 ** 2 + col3 * 10)
<ipython-input-39-c220950852c4>:2: RuntimeWarning: invalid value encountered in multiply
  return np.sqrt(col1 ** 3 + col2 ** 2 + col3 * 10)
<ipython-input-39-c220950852c4>:2: RuntimeWarning: invalid value encountered in sqrt
  return np.sqrt(col1 ** 3 + col2 ** 2 + col3 * 10)


# The eval the function of Pandas

There are two versions -
> pd.eval (higher-level)

> df.eval (in the context of DataFrames).

**Like Numba, you should have at least +10,000 samples in the DataFrame to see improvements. But once you do, you will see sizeable benefits in speed.**

Let’s run our crazy_function in the context of df.eval:

In [45]:
%%time

massive_df.eval("f1001 = (f1 ** 3 + f56 ** 2 + f44 * 10) ** 0.5", inplace=True)

CPU times: user 211 ms, sys: 77 ms, total: 288 ms
Wall time: 224 ms


It's not as fast as vectorization or Numba, but it has several benefits. First, you write much less code by avoiding references to the DataFrame name. Next, it significantly speeds up non-math operations on DataFrames like boolean indexing, comparisons, and many more.



**Even though we have 1-million row dataset, all our operations were under a single second.**



### **When you are not doing mathematical manipulation, evaluate your statements in pd.eval.**